### PR TITLE
feat: implement query search 

### LINKED ISSUE #646: Major feature: Support query based search
Story
As a user, I want to go to the exact query cell when searching queries. I also want to check out past executions by me or others.

Requirements:
- We should support filtering by
    - environment (always applied)
        - Note: for adhoc queries, the environment is a multi value field from its associated query engine
    - authors (by default, all users)
    - query engines (by default, all engines in environment)
    - query cell or query execution (by default, both)
    - query type ('SELECT', 'INSERT', 'CREATE', 'DELETE') (by default
    - startTime
    - duration
    - tables in query (we can index by fully qualified name)
- We should support ordering results by
    - recency (default)
    - relevancy
- Once query search is implemented, this should be the first tab in the cmd+k search
- Query search by execution should also be supported in the executions sidebar tab
- Only successful query executions are indexed, however this should be flexible for customization
- We update the search index on the following events
     - update_data_cell
     - delete_data_cell (removal)
     - query_execution_completion
     - query_execution_deletion (removal)
     - Note:  delete_data_cell, query_execution_deletion may be a bulk action such as when removing an archived datadoc or removing all query executions past a certain date. Please ensure the index would update in those cases.

Suggested order for implementation (Separate PR recommended):
- Define the data model in elasticsearch
- Create the pipeline that syncs the data to es
- API to perform query search and the associated cmd+k search UI
- Update to query execution sidebar

### PR DESCRIPTION
**Pull request by kgopal492**
closes #646 

* Add query search by adding indices for query cells and query executions in elasticsearch
* Add options to filter by:
  * author
  * query engine
  * tables used in query
  * query type (cell or execution)
  * statement type (select, insert, delete, etc)
  * creation time
  * duration

* Query Cells:
  * `data_cell` of type query are indexed as query cells when the data cell is created or updated
  * only query cells on active (not archived) data docs are indexed
  * deleted when the data doc has been archived or cell has been deleted
 
* Query Executions:
  *  only successful `query_execution`s are indexed
  * deleted when the `clean_up_query_execution` job runs

* Search icon (link to query search) is added to the execution sidebar
Screen Shot 2021-12-22 at 1 40 54 PM

* Query search filters
Screen Shot 2021-12-22 at 1 36 57 PM

* Table filter UI for multiple metastores
Screen Shot 2021-12-22 at 1 41 58 PM

* Query search page
Screen Shot 2021-12-22 at 3 40 31 PM
Screen Shot 2021-12-22 at 3 40 37 PM
  * query result title is:
    * for query cells: title of query cell or "Untitled"
    * for query executions: "{query cell title} > Execution {id}" for data cell executions or "Adoc Execution {id}"

* Query search page with keyword search entered
Screen Shot 2021-12-22 at 3 42 45 PM

* Updated Docs

Breaking Changes
Screen Shot 2022-01-25 at 12 34 28 PM

Re-Initialize ES
Screen Shot 2022-01-25 at 12 34 48 PM

-------
Update:

* By default, query search is not performed if user hasn't entered search string/filters
Screen Shot 2022-02-16 at 5 21 00 PM

* User can toggle between matched results highlighting & themed code highlight

https://user-images.githubusercontent.com/23270317/154390769-bd0a8f0e-b643-4650-b762-1b742754f6e6.mov

### CODE REVIEW

File: querybook/webapp/components/Search/SearchOverview.tsx [hunk 1]
```diff
@@ -320,73 +367,263 @@ export const SearchOverview: React.FunctionComponent = () => {
...
+                    value: engine.id,
+                }))}
+                isClearable
+            />
+        </div>
+    );
+
+    const queryTypeFilterDOM = (
+        <div className="filter-query-type">
+            <ReactSelect
```
Comment (czgu) [inline querybook/webapp/components/Search/SearchOverview.tsx:422]: use SimpleReactSelect

File: querybook/webapp/components/Search/SearchOverview.tsx [hunk 2]
```diff
@@ -320,73 +367,263 @@ export const SearchOverview: React.FunctionComponent = () => {
...
+                isMulti
+                isClearable
+            />
+        </div>
+    );
+
+    const tableFilterDOM = (
+        <TableSelect
+            tableNames={searchFilters['full_table_name'] || []}
+            onSelect={(tableName) => {
```
Comment (czgu) [inline querybook/webapp/components/Search/SearchOverview.tsx:476]: let TableSelect manage the state of tableNames, so onSelect can just be `(newTableNames: string[]) => updateSearchFilter('full_table_name', newTableNames)`

File: querybook/webapp/components/Search/SearchOverview.tsx [hunk 3]
```diff
@@ -320,73 +367,263 @@ export const SearchOverview: React.FunctionComponent = () => {
...
+                updateSearchFilter('full_table_name', newTableNames);
+            }}
+            selectProps={{
+                autoFocus: true,
+            }}
+            clearAfterSelect
+        />
+    );
+
+    const getAuthorFiltersDOM = React.useCallback(
```
Comment (czgu) [inline querybook/webapp/components/Search/SearchOverview.tsx:501]: no need to memo DOM renders

File: querybook/webapp/components/Search/SearchResultItem.tsx [hunk 4]
```diff
@@ -47,6 +52,93 @@ function openClick(
...
+        title,
+    } = preview;
+    const { userInfo: authorInfo, loading } = useUser({ uid: authorUid });
+
+    const isQueryCell = preview.query_type === 'query_cell';
+
+    const url = isQueryCell
+        ? `/${environmentName}/datadoc/${preview.data_doc_id}/?cellId=${id}`
+        : `/${environmentName}/query_execution/${id}/`;
+    const handleClick = React.useMemo(() => openClick.bind(null, url), [url]);
```
Comment (czgu) [inline querybook/webapp/components/Search/SearchResultItem.tsx:81]: `useCallback(() => openClick(url), [url])`

File: querybook/webapp/components/Search/SearchResultItem.tsx [hunk 5]
```diff
@@ -47,6 +52,93 @@ function openClick(
...
+
+    return (
+        <div className="SearchResultItem QueryItem" onClick={handleClick}>
+            <div className="result-items">
+                <a className="result-items-top horizontal-space-between">
+                    <HighlightTitle
+                        title={resultTitle}
+                        searchString={searchString}
+                    />
+                    <Tag>{queryEngine?.name}</Tag>
```
Comment (czgu) [inline querybook/webapp/components/Search/SearchResultItem.tsx:118]: `queryEngine ? <Tag>...</Tag> : null`

File: querybook/webapp/components/Search/SearchResultItem.tsx [hunk 6]
```diff
@@ -47,6 +52,93 @@ function openClick(
...
+        return <div className="SearchResultItem QueryItem">Loading...</div>;
+    }
+
+    // Query cell title is data cell title
+    // Query execution title is "<title> > Execution <id>" for data cell executions
+    // or "Adhoc Execution <id>" for adhoc exeuctions
+    const resultTitle = isQueryCell
+        ? title ?? 'Untitled'
+        : `${title != null ? `${title} >` : 'Adhoc'} Execution ${id}`;
+    const queryTextHighlightedContent = (preview.highlight || {}).query_text;
```
Comment (czgu) [inline querybook/webapp/components/Search/SearchResultItem.tsx:94]: preview.highlight?.query_text

File: querybook/webapp/components/Search/SearchResultItem.tsx [hunk 7]
```diff
@@ -47,6 +52,93 @@ function openClick(
...
+            className="result-item-description"
+            dangerouslySetInnerHTML={{
+                __html: formatHighlightStrings(queryTextContent),
+            }}
+        />
+    );
+
+    const queryEngine = queryEngineById[engineId];
+    const language = queryEngine?.language ?? 'presto';
+    const formattedQuery = format(queryText, language, {
```
Comment (czgu) [inline querybook/webapp/components/Search/SearchResultItem.tsx:106]: formatting per render is kinda expensive,  i dont think its needed for search. If you want to keep it can you make ThemedCodeHighlight support auto formatting?

File: querybook/webapp/components/Search/TableSelect.tsx [hunk 8]
```diff
@@ -0,0 +1,130 @@
...
+import { useSelector } from 'react-redux';
+import { queryMetastoresSelector } from 'redux/dataSources/selector';
+import { IStoreState } from 'redux/store/types';
+import { HoverIconTag } from 'ui/Tag/Tag';
+
+import './TableSelect.scss';
+
+interface ITableSelectProps {
+    tableNames: string[];
+    onSelect: (tableName: string) => any;
```
Comment (czgu) [inline querybook/webapp/components/Search/TableSelect.tsx:18]: use => void if you don't expect anything to be returned

File: querybook/webapp/components/Search/TableSelect.tsx [hunk 9]
```diff
@@ -0,0 +1,130 @@
...
+    const tableReactSelectStyle = makeReactSelectStyle(usePortalMenu);
+    if (usePortalMenu) {
+        asyncSelectProps.menuPortalTarget = overlayRoot;
+    }
+    if (clearAfterSelect) {
+        asyncSelectProps.value = null;
+    }
+
+    const loadOptions = debounce(
+        (tableName, callback) => {
```
Comment (czgu) [inline querybook/webapp/components/Search/TableSelect.tsx:52]: instead of callback, use promise or async

File: querybook/webapp/redux/search/action.ts [hunk 10]
```diff
@@ -99,20 +103,30 @@ export function performSearch(): ThunkResult<Promise<ISearchPreview[]>> {
...
-                              state.environment.currentEnvironmentId,
-                      })
-                    : SearchTableResource.search({
-                          ...searchParams,
-                          metastore_id:
-                              state.dataTableSearch.metastoreId ||
-                              queryMetastoresSelector(state)[0].id,
-                          fields: Object.keys(searchState.searchFields),
-                      });
+            let searchRequest;
```
Comment (czgu) [inline querybook/webapp/redux/search/action.ts:106]: can you ensure the typing works in this case?

File: querybook/webapp/ui/SearchBar/SearchBar.tsx [hunk 11]
```diff
@@ -46,7 +46,7 @@ export const SearchBar: React.FunctionComponent<ISearchBarProps> = ({
             // setTimeout is needed to make the focus work
             // otherwise the component
             setTimeout(() => {
-                inputRef.current.focus();
+                inputRef?.current?.focus();
```
Comment (czgu) [inline querybook/webapp/ui/SearchBar/SearchBar.tsx:49]: inputRef.current can be null?
in this case, inputRef shouldn't be null, can you change it to `inputRef.current?.focus()`

File: querybook/webapp/components/Search/SearchOverview.tsx (hunk 2 above)
Comment (kgopal492) [reply querybook/webapp/components/Search/SearchOverview.tsx:476]: updated to `setTableNames` so that the TableSelect component can also use that in `onRemove`

File: querybook/webapp/components/Search/SearchResultItem.tsx (hunk 7 above)
Comment (kgopal492) [reply querybook/webapp/components/Search/SearchResultItem.tsx:106]: removed it

File: querybook/webapp/components/Search/SearchResultItem.tsx (hunk 4 above)
Comment (kgopal492) [reply querybook/webapp/components/Search/SearchResultItem.tsx:81]: Added event to function call as well to make sure it handles command click correctly

File: querybook/server/logic/query_execution.py [hunk 12]
```diff
@@ -519,6 +521,42 @@ def clean_up_query_execution(dry_run=False, session=None):
...
+
+
+@with_session
+def get_successful_adhoc_query_executions(offset=0, limit=100, session=None):
+    return (
+        session.query(QueryExecution)
+        .filter(QueryExecution.status == QueryExecutionStatus.DONE)
+        .join(DataCellQueryExecution, isouter=True)
+        .filter(DataCellQueryExecution.id.is_(None))
+        .order_by(QueryExecution.id.desc())
```
[the remark this answers is not in the document]
Comment (kgopal492) [reply querybook/server/logic/query_execution.py:538]: removed order by

File: querybook/server/logic/datadoc_collab.py [hunk 13]
```diff
@@ -177,7 +177,7 @@ def update_data_cell(cell_id, fields, sid="", session=None):
     assert_can_write(data_doc.id, session=session)
     verify_environment_permission([data_doc.environment_id])
     session.commit()
-
+    logic.update_es_query_cell_by_id(data_cell.id)
```
Comment (czgu) [inline querybook/server/logic/datadoc_collab.py:180]: hm i guess the logic is a bit convoluted now, lets do `assert_can_write` and `verify_environment_permission` first before update_data_cell and have the es update logic there

File: querybook/server/logic/datadoc.py [hunk 14]
```diff
@@ -894,5 +914,60 @@ def get_data_cells_executions(ids, session=None):
...
+
+@with_session
+def get_all_query_cells(offset=0, limit=100, session=None):
+    return (
+        session.query(DataCell)
+        .filter_by(cell_type=DataCellType.query)
+        .join(DataDocDataCell)
+        .join(DataDoc)
+        .filter(DataDoc.archived.is_(False))
+        .order_by(DataCell.id.desc())
```
Comment (czgu) [inline querybook/server/logic/datadoc.py:960]: unnecessary order by

File: querybook/server/tasks/run_query.py [hunk 15]
```diff
@@ -56,6 +56,7 @@ def run_query_task(self, query_execution_id):
                 query_execution_id, executor, error_message, session=session
             )
             notifiy_on_execution_completion(query_execution_id, session=session)
+            qe_logic.update_es_query_execution_by_id(query_execution_id)
```
Comment (czgu) [inline querybook/server/tasks/run_query.py:59]: given you are already in a celery task, you can call the logic directly to update instead of spawning a new one

File: querybook/webapp/components/Search/SearchOverview.tsx [hunk 16]
```diff
@@ -320,8 +363,117 @@ export const SearchOverview: React.FunctionComponent = () => {
...
+        </div>
+    );
+
+    const queryEngineFilterDOM = (
+        <div className="filter-query-engine">
+            <ReactSelect
+                styles={defaultReactSelectStyles}
+                value={
+                    searchFilters['engine_id'] && {
+                        label: queryEngineById[searchFilters['engine_id']].name,
```
Comment (czgu) [inline querybook/webapp/components/Search/SearchOverview.tsx:397]: queryEngineById[searchFilters['engine_id']]?.name

File: querybook/webapp/components/Search/SearchResultItem.tsx [hunk 17]
```diff
@@ -118,7 +205,11 @@ export const DataTableItem: React.FunctionComponent<IDataTableItemProps> = ({
         name,
         schema,
     } = preview;
-    const handleClick = React.useMemo(() => openClick.bind(null, url), [url]);
+    const handleClick = useCallback(
```
Comment (czgu) [inline querybook/webapp/components/Search/SearchResultItem.tsx:208]: you are right about using useMemo, can you revert the change back 🙏

File: requirements/base.txt [hunk 18]
```diff
@@ -31,7 +31,7 @@ pyyaml==5.4.1
 sqlalchemy==1.3.17
 pymysql==0.9.3
 requests==2.22.0
-elasticsearch==7.13.4
+elasticsearch==7.16.2
```
Comment (czgu) [inline requirements/base.txt:34]: please revert back this change

File: querybook/webapp/components/Search/TableSelect.tsx [hunk 19]
```diff
@@ -0,0 +1,128 @@
...
+import { useSelector } from 'react-redux';
+import { queryMetastoresSelector } from 'redux/dataSources/selector';
+import { IStoreState } from 'redux/store/types';
+import { HoverIconTag } from 'ui/Tag/Tag';
+
+import './TableSelect.scss';
+
+interface ITableSelectProps {
+    tableNames: string[];
+    setTableNames: (tableNames: string[]) => void;
```
Comment (czgu) [inline querybook/webapp/components/Search/TableSelect.tsx:17]: onTableNamesChange

File: querybook/webapp/ui/NumberInput/NumberInput.tsx [hunk 20]
```diff
@@ -0,0 +1,18 @@
+import React from 'react';
+
+interface INumberInputProps extends React.HTMLProps<HTMLInputElement> {
+    setValue: (value: number | null) => void;
```
Comment (czgu) [inline querybook/webapp/ui/NumberInput/NumberInput.tsx:4]: use onChange

**Review by czgu — COMMENTED** [review]
can you update the version to be 3.1.0 and add upgrade instructions on https://www.querybook.org/docs/changelog?

File: querybook/server/logic/datadoc.py [hunk 21]
```diff
@@ -66,9 +69,12 @@ def create_data_doc(
             commit=False,
             session=session,
         )
+        data_cell_ids.append(data_cell.id)
     if commit:
         session.commit()
         update_es_data_doc_by_id(data_doc.id)
+        for cell_id in data_cell_ids:
+            update_es_query_cell_by_id(cell_id)
```
[the remark this answers is not in the document]
Comment (kgopal492) [reply querybook/server/logic/datadoc.py:77]: Using `update_es_query_cells_by_data_doc_id` instead to get only the query cells

**Review by czgu — APPROVED** [review]
Great job!

File: querybook/webapp/redux/search/reducer.ts [hunk 22]
```diff
@@ -19,7 +19,7 @@ const initialPaginationState: ISearchPaginationState = {
 const initialSearchParamState = {
     searchFilters: {},
     searchFields: {},
-    searchOrder: SearchOrder.Relevance,
+    searchOrder: SearchOrder.Recency,
```
Comment (czgu) [inline querybook/webapp/redux/search/reducer.ts:22]: still default to Relevance?

**Review by czgu — APPROVED** [review]

File: querybook/server/lib/query_analysis/lineage.py [hunk 23]
```diff
@@ -103,7 +106,10 @@ def get_statement_placeholders(statement):
             placeholders.append(token.get_real_name())
         elif token and isinstance(token, sqlparse.sql.IdentifierList):
             for identifier in token.get_identifiers():
-                placeholders.append(identifier.get_real_name())
+                try:
```
[the remark this answers is not in the document]
Comment (kgopal492) [reply querybook/server/lib/query_analysis/lineage.py]: Did a bit of digging - apparently the `get_identifiers` function returns some objects that aren't `sqlparse.sql.Identifier`s. Added a check to make sure the token is an identifier.
